### DISCUSSION TITLE
Custom Functions - Part 2

### DISCUSSION THREAD

**Discussion by lvca** [ArcadeData/arcadedb#1155] [Ideas]
Part 1: https://github.com/ArcadeData/arcadedb/discussions/281

ArcadeDB already supports custom functions written in any language supported by the JVM. The most common is Javascript and it's available out of the box.

In order to define a new function "math" in Javascript you can execute this (This syntax has been inherited from OrientDB):

```sql
define function math.sum \"return a + b\" parameters [a,b] language js
```

Since functions are not persistent, every time a server starts, the function must be registered. Also, there is no easy way to write and run functions from Studio or an IDE. 

I believe custom functions will play an important role in the future of ArcadeDB, mainly because:
- it's hard to write a general-purpose query optimizer on graphs
- complex queries are hard for beginners because they require a deep knowledge of ArcadeDB (or Cypher)
- complex queries are hard to debug

What if ArcadeDB provides the support of IDEs and a minimal code editor in Studio? Also, if possible, the ability to debug the code in the server would be extremely helpful for the developer.

## IDE Support

In order to use any IDE with ArcadeDB functions, the easiest way would be to have a new folder "functions" under the database directory containing the code of the functions in files with the suffix `.js` for Javascript. In this way, you can work with any IDE. A  file `database.js` could be created by default containing all the functions (with no implementation) of the supported functions callable from ArcadeDB. Example:

database.js

```js
function database_lookupDocumentByRID( databaseInstance, rid ) {
}
```

The IDE could load this file and provide a syntax completion out of the box. Of course, this file must be updated with the real server-side functions available on the Server.

customer.js

```js
function getById( customerId ) {
  let result = database_lookupDocumentByKey( db, "Customer", ["id"], [ customerId ]  );
  return result.length > 0 ? result[0] : null;
}
```

The provided JS functions could mimic the current Java API with the class name as a prefix separated by an underscore. This would avoid using OO in Javascript.

At startup, the ArcadeDB server will load all the files with extensions `.js` from the `functions` directory and registers the functions. The functions above would be translated as:

```sql
define function customer.getById
"let result = database_lookupDocumentByKey( db, \"Customer\", [\"id\"], [ customerId ]  ); return result.length > 0 ? result[0] : null;" 
parameters [customerId] language js
```

Note that the file name becomes the library (customer) where the function is registered.

And the user can call the function from SQL (console, studio, postgres driver, etc), such as:

```sql
select `customer.getById( 12345 )` as customer
```

Or SQL from Java with a prepared statement:

```java
final ResultSet result = database.command("sql", "select `customer.getById(?)` as customer", 12345);
```

Or directly via Java API:

```java
Vertex customer = database.getSchema().getFunction("customer", "getById").execute(12345);
```

Thoughts? Ideas? WDYT?

Contributions are welcome!

**Comment by mdre** [comment]
It a great Idea. We have a lot of functions defined in OrientDB that must be ported. 
Consider to do the same but with a Java. We have a FullPath function that returns all the path found between two vertex. Now I hack the code and add the functions direct inside ArcadeDB but it is not a reccomended way.

**Reply by lvca** [reply]
@mdre how did you write such functions with OrientDB? Java classes you put under the lib directory of OrientDB?

**Reply by mdre** [reply]
We put it in the plugins directory.
Here is the repo:

https://github.com/mdre/FullPath

**Reply by lvca** [reply]
Thanks @mdre, out of this scope, I'm considering adding your function to ArcadeDB's graph library. What's the graph algorithm you have used to find all the paths?

**Comment by tolgaulas** [comment]
@lvca , should you take the path of "functions" under database folder for script files, would you pls consider making it either a variable that we can set anywhere we like or functionable with symbolic link so we can point from wherever we like ( we keep things under the project repos. With orientdb we have setup_db script which init a db along with the existing embedded functions; then we can store our functions in similar fashion.)

**Reply by lvca** [reply]
a symbolic link would work, but having in a separate settings it's something we can also easily do ;-)

**Comment by ansarizafar** [comment]
I like this idea. Is there any update on this feature?

**Comment by mdre** [comment]
I think it's a great idea.
Sorry but I was full time on other projects. I'am  nearly  starting again
with Arcade.
To answer Luca's question, I implemented the OServerPluginAbstract and put
it in the lib.
If you can mimic the same interface it could be great to have full speed in
functions.
See you soon.

**Comment by mdre** [comment]
There's no algorithm because it just a full path. I only do a deep
traverse till to node or max deep and continue with backtracking from
the last node. It is really important to use the filter with the
include/exclude filters to limit the function.

The git repo is this: https://github.com/mdre/FullPath

**Reply by lvca** [reply]
Do you take care of circular references?

**Comment by mdre** [comment]
Yes. I prevent finding if a node exists in the current path before it
is added. If it exists it is ignored because it is processing in the
backtracking.
When a path is found, it is added to the resulting collection of paths.
